jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -1074,204 +950,119 @@ ecma_number_to_utf8_string (ecma_number_t num, /**< ecma-number */
+  }
 
-    if (ecma_number_is_zero (num))
-    {
-      // 2.
-      *dst_p++ = digits[0];
+  dst_p = buffer_p;
 
-      JERRY_ASSERT (dst_p <= buffer_p + buffer_size);
-      size = (lit_utf8_size_t) (dst_p - buffer_p);
-    }
-    else if (ecma_number_is_negative (num))
-    {
-      // 3.
-      *dst_p++ = minus_char;
-      lit_utf8_size_t new_buffer_size = (lit_utf8_size_t) ((buffer_p + buffer_size) - dst_p);
-      size = 1 + ecma_number_to_utf8_string (ecma_number_negate (num), dst_p, new_buffer_size);
-    }
-    else if (ecma_number_is_infinity (num))
-    {
-      // 4.
-      dst_p = lit_copy_magic_string_to_buffer (LIT_MAGIC_STRING_INFINITY_UL, buffer_p, buffer_size);
-      size = (lit_utf8_size_t) (dst_p - buffer_p);
-    }
-    else
-    {
-      ecma_number_t p = ecma_number_get_prev (num);
-      ecma_number_t q = ecma_number_get_next (p);
-      JERRY_ASSERT (q == num);
+  if (ecma_number_is_negative (num))
+  {
+    // 3.
+    *dst_p++ = '-';
+    num = ecma_number_negate (num);
+  }
 
-      // 5.
-      uint32_t num_uint32 = ecma_number_to_uint32 (num);
+  if (ecma_number_is_infinity (num))
+  {
+    // 4.
+    dst_p = lit_copy_magic_string_to_buffer (LIT_MAGIC_STRING_INFINITY_UL, dst_p,
+                                             (lit_utf8_size_t) (buffer_p + buffer_size - dst_p));
+    JERRY_ASSERT (dst_p <= buffer_p + buffer_size);
+    return (lit_utf8_size_t) (dst_p - buffer_p);
+  }
 
-      if (((ecma_number_t) num_uint32) == num)
-      {
-        size = ecma_uint32_to_utf8_string (num_uint32, dst_p, buffer_size);
-      }
-      else
-      {
-        /* mantissa */
-        uint64_t s;
-        /* decimal exponent */
-        int32_t n;
-        /* number of digits in k */
-        int32_t k;
+  JERRY_ASSERT (ecma_number_get_next (ecma_number_get_prev (num)) == num);
 
-        ecma_number_to_decimal (num, &s, &k, &n);
+  // 5.
+  uint32_t num_uint32 = ecma_number_to_uint32 (num);
 
-        // 6.
-        if (k <= n && n <= 21)
-        {
-          dst_p += n;
-          JERRY_ASSERT (dst_p <= buffer_p + buffer_size);
+  if (((ecma_number_t) num_uint32) == num)
+  {
+    dst_p += ecma_uint32_to_utf8_string (num_uint32, dst_p, (lit_utf8_size_t) (buffer_p + buffer_size - dst_p));
+    JERRY_ASSERT (dst_p <= buffer_p + buffer_size);
+    return (lit_utf8_size_t) (dst_p - buffer_p);
+  }
 
-          size = (lit_utf8_size_t) (dst_p - buffer_p);
+  /* decimal exponent */
+  int32_t n;
+  /* number of digits in mantissa */
+  int32_t k;
 
-          for (int32_t i = 0; i < n - k; i++)
-          {
-            *--dst_p = digits[0];
-          }
+  k = (int32_t) ecma_number_to_decimal (num, dst_p, &n);
 
-          for (int32_t i = 0; i < k; i++)
-          {
-            *--dst_p = digits[s % 10];
-            s /= 10;
-          }
-        }
-        else if (0 < n && n <= 21)
-        {
-          // 7.
-          dst_p += k + 1;
-          JERRY_ASSERT (dst_p <= buffer_p + buffer_size);
+  if (k <= n && n <= 21)
```

Comment:
From the standard. As the comments explain (of the function and of the if body), it's algorithm of section 9.8.1, step 6. For the sake of completeness: "6. If k <= n <= 21, return the String consisting of the k digits of the decimal representation of s (in order, with no leading zeroes), followed by n-k occurrences of the character '0‘."